### PR TITLE
[core][SDPA] Allow a quantized key or value operand

### PR DESCRIPTION
**Pull request by dmenig**
### Details:

 - `ScaledDotProductAttention::validate_and_infer_types()` merges every operand into the query's
   element type, so an `i8` key is rejected in core with `Mixed input types are not supported`.
   A core op validates when it is constructed, so this happens while the model is being built or
   read — before any plugin is consulted. A plugin-side matcher cannot help a model whose IR
   already carries a quantized key, because the graph cannot be constructed at all.
 - This admits `i8`, `u8`, `i4` and `u4` on the **key and value operands only**. Unchanged: the
   output type still follows the query; a quantized query is still rejected as non-floating-point;
   `i32` is still rejected as a mixed type; and the mask, scale and sink operands still take part
   in the merge exactly as before. Shape inference is untouched.
 - The rule is single-sourced as `ScaledDotProductAttention::is_quantized_kv_type()` and
   `has_quantized_kv()`, so a consumer cannot drift from the set the op accepts.

**Why not `Convert` + `Multiply` in the IR instead?** That form is expressible today and is the
right answer whenever the dequantization is a graph operation: a plugin matcher folds it into its
own attention primitive and the `Convert` never runs. It is not the case this addresses. What is
missing is the ability for a model to state that the key *is* a code tensor, where the producer
already emits codes and no widening node exists to fold. Today such a model cannot be constructed,
so the question cannot even be put to a plugin.

**What this does not add, and the risk that carries.** The operand has no scale and no zero point,
so the mapping from codes to values is defined by the consuming implementation, not by the IR.
That is a real limitation. Its failure mode is not a compile error but a plausible wrong number,
because a consumer that reads the operand at the query's precision widens the codes as magnitudes.
`intel_gpu::op::SDPA` is **not** a precedent for leaving the mapping out: when its key and value
are quantized it always carries 2 to 4 companion scale/zero-point operands, so it is
self-describing in a way this op is not.

**The alternative is to give the public op optional `key_scale`/`value_scale` (and zero-point)
operands.** That makes the IR self-describing and turns a missed consumer into an arity error
instead of a wrong answer, at the cost of a larger change to a released opset. I would rather
agree that with a maintainer than guess — **if it is the preferred direction, say so and I will
rework this into it.**

Meanwhile every in-tree consumer that reads these operands at the query's precision now declines
or reports by name, instead of reading the codes:

| consumer | without the guard | with it |
|---|---|---|
| `ScaledDotProductAttentionDecomposition` | emits MatMuls; the type error surfaces on a synthesized node that exists in no model | rejects, naming the SDPA node and the type |
| `StatefulSDPAFusion` (CPU) | fuses, then widens the codes as magnitudes | declines; the decomposition then reports |
| `SDPAFuseTransposeReshape` (CPU x64) | fuses, then widens the codes as magnitudes | declines; the decomposition then reports |
| `ScaledDotProductAttentionDecomposition` callback (GPU) | keeps the op, and `SDPAOpt` dispatches on an `i8` key with no compression operands | decomposes, so the rejection is reported |
| `SDPAToPagedAttention::run_on_model` | rewrites into `PagedAttentionExtension`, whose key/value type list is empty | rejects by name before any layer is rewritten or any parameter added — this is an explicit conversion request, so declining silently would leave a half-converted model, and rejecting from inside the matcher callback (the first version of this PR) left a model with a supported layer ahead of a quantized one half-converted too |
| `SDPAToVLSDPA` | rewrites into VLSDPA, whose own validation then throws `Mixed input types of K/V are not supported` — already a clean error, unlike the rows above | declines, so the rejection names the SDPA node |
| template backend reference | trips a `Tensor::data<T>` assert on the first `infer()` | names the unsupported operand, still at inference — op support there is expressed as `evaluate_node`, so it cannot reject at compile time |

**Known gap, stated deliberately:** NPUW carries its own copies of the SDPA decomposition
(`npuw_transformations/optimize_value_tensors.cpp`, `whisper/prepare_whisper_model.cpp`) which are
independent of `ov::pass::ScaledDotProductAttentionDecomposition` and do not gain this diagnostic.
They predate this change; I have not touched them, and I would rather flag that than imply the
sweep is complete.

 - The specification gains a *T_QUANT* type for the two operands, with a note that the mapping from
   codes to values is defined by the consuming implementation and that an implementation without
   such a mapping is expected to reject the op.

**On measurement:** there is nothing to benchmark — the change emits no code. It makes a
quantized-operand attention path expressible on the public op; the speed belongs to whichever
kernel takes it up and should be measured there. I am happy to produce `benchmark_app` numbers on
whichever public quantized-attention model the maintainers consider representative.

**Tests added:**
 - `type_prop`: quantized key; quantized key and value on the 5- and 6-input forms; dynamic query
   type with a quantized key.
 - `type_prop` negatives proving the relaxation is narrow: quantized query, `i32` key, quantized
   mask, quantized sink, quantized scale — each still rejected with its original message.
 - `ScaledDotProductAttentionDecomposition` rejects a quantized key and a quantized value with a
   message naming the SDPA node.
 - `SDPAToVLSDPA` leaves a quantized-key model unconverted.
 - IR serialization round-trip: a `u4` key and `i8` value survive `Serialize` and re-read with
   their types intact, which is the case this change exists for.
 - `SDPAToPAQuantizedKVTest.RejectedBeforeAnyLayerIsConverted`: two stacked OPT-125m attention
   layers, the first convertible and the second with an `i8` K/V cache. `SDPAToPagedAttention`
   throws and leaves both SDPAs in place with no `PagedAttentionExtension` built. Against the
   in-callback guard this fails with one layer converted, which is the half-converted model.
 - `StateConcatSDPAQuantizedKVShouldNotFuse` in `state_concat_sdpa.cpp`: a stateful `i8` K/V
   cache graph run through `StatefulSDPAFusion` compares equal to the untouched graph, so the
   original SDPA survives and no `ScaledDotProductAttentionWithKVCache` is constructed.

### Tickets:
 - N/A

### AI Assistance:
 - AI assistance used: yes
 - Claude Code drafted the change, the tests and this description. Human validation performed:
   built and ran `ov_core_unit_tests` (13824 passed / 8 skipped / 0 failed), `ov_transformations_tests` (10182 passed / 0 failed) and
   `ov_cpu_unit_tests` (2672 passed / 1 failed, the failure being ModelPreferThreadsIntegrationTest.x86_hybrid_fp32_main_cores_only, a host-topology check this change does not touch) in full from this branch at `b3b77e0628`, and built the CPU and GPU
   plugins to compile-check the guards that no unit test reaches. The new type checks
   were mutation-tested: forcing the predicate to accept nothing fails the four positive tests,
   and forcing it to accept everything fails the `i32`-key negative test and an existing upstream
   boolean-mask test, confirming the allow-list has to be exactly this wide.
 - Guard coverage, stated plainly: the decomposition, `SDPAToVLSDPA`, `SDPAToPagedAttention` and
   `StatefulSDPAFusion` guards have unit tests, and the shared predicate is pinned by its own test.
   The two new tests were mutation-checked: restoring the in-callback PagedAttention guard fails
   `RejectedBeforeAnyLayerIsConverted` and only it; disabling the `StatefulSDPAFusion` guard fails
   `StateConcatSDPAQuantizedKVShouldNotFuse` and only it. The `SDPAFuseTransposeReshape`,
   GPU-callback and template-reference guards still have **no automated test** — they were verified
   by hand by compiling a matching model and observing the error. Plugin functional tests would be
   the durable place for the first two; I have not added them.
 - `clang-format` 18 (the version `cmake/developer_package/clang_format/clang_format.cmake` pins)
   is clean on every changed file that a clang-format target governs. The nine lines added to
   `intel_gpu/src/plugin/transformations_pipeline.cpp` are clang-format-18-clean, but that file
   carries pre-existing formatting drift (identical on the base commit) and `intel_gpu` has no
   clang-format target, so I have not reformatted it.

### CODE REVIEW

**Comment by p-durandin** [comment]
build_jenkins

File: docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.rst [hunk 1]
```diff
@@ -74,6 +74,14 @@ omitting training-related parameter.
 
     The sink input is available since 2025.4 OpenVINO release.
 
+.. note::
+
+    A *T_QUANT* ``key`` or ``value`` is available since 2026.4 OpenVINO release. The pseudo-code
+    above describes the operation for a *T* ``key`` and ``value`` only. When either is *T_QUANT*
+    it holds quantization codes rather than values, and the mapping from codes to values is
+    defined by the implementation that consumes the operation, not by this specification. An
+    implementation that does not define such a mapping is expected to reject the operation.
```
[the remark this answers is not in the document]
Comment (dmenig) [reply docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.rst:83]: > Not doing this, and the reason is that the operands have nowhere to go and nothing to mean.
> 
> This op's optional inputs are positional: attention_mask at 3, scale at 4, sink at 5. Scale and zero-point operands would have to sit at 6 and beyond, so no model could supply a key scale without also supplying a sink it does not have. And giving those operands a defined code-to-value mapping makes quantized attention a semantic that every implementation owes — the reference and template backends included — which is a far larger change to a released opset than the type relaxation it would be replacing.
> 
> The self-describing form you are describing already exists one level down, which is where it belongs. KVCacheCompression in the GPU plugin matches the internal IndirectSDPA and rebuilds it carrying compression scales and zero points as real operands. An implementation that can honour a mapping states it there; the public op deliberately does not.
> 
> What this PR does instead is make the ambiguity impossible to hit silently. The specification says an implementation without a code-to-value mapping is expected to reject the op, all seven in-tree consumers now decline or report the node by name rather than reading codes at the query's precision, and a serialization round-trip test pins that a u4 key and i8 value survive Serialize with their types intact. The residual exposure is an out-of-tree consumer that ignores the spec text, and that is the trade I am making knowingly.
> 
>

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp [hunk 2]
```diff
@@ -502,6 +502,18 @@ ov::pass::StateManagementPattern::StateManagementPattern(PaParams& pa_params,
     auto sdpa_variants = std::make_shared<Or>(OutputVector{sdpa_with_4_inputs, sdpa_with_5_inputs, sdpa_with_6_inputs});
 
     ov::matcher_pass_callback callback = [=, &pa_params, &results, &var_ids_to_remove](Matcher& m) {
+        // PagedAttentionExtension has no operand able to carry the mapping from quantization
+        // codes to values. SDPAToPagedAttention is an explicit conversion request, so report it
+        // against the node rather than leaving a half-converted model behind.
+        const auto sdpa_root = ov::as_type_ptr<v13::ScaledDotProductAttention>(m.get_match_root());
+        NODE_VALIDATION_CHECK(sdpa_root.get(),
+                              !v13::ScaledDotProductAttention::has_quantized_kv(*sdpa_root),
```
[the remark this answers is not in the document]
Comment (dmenig) [reply src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: > Agreed, and fixed. The check is gone from the matcher callback and is now a pre-scan at the top of `SDPAToPagedAttention::run_on_model`, alongside the assertions that already run there before anything is mutated.
>
> The regression test is `RejectedBeforeAnyLayerIsConverted`, built on the existing OPT-125m model builders: two stacked attention layers, the first convertible and the second carrying an i8 K/V cache. Against the previous code it fails with one `PagedAttentionExtension` and one surviving SDPA — exactly the half-converted model you describe. With the pre-scan both layers are untouched.
>
> I checked that it fails against the old guard rather than assuming it would. My first attempt at this test passed either way and did not test the fix at all: the two attention layers were built side by side, so the rewrite reached the quantized one first and never converted anything. Stacking them the way a transformer actually is exposed it.
>
> `ov_core_unit_tests` 13804 passed / 0 failed and `ov_transformations_tests` 10151 passed / 0 failed with the change in. Landed in b3b77e0628, as a separate commit on top of the reviewed head.

File: src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/stateful_sdpa_fusion.cpp [hunk 3]
```diff
@@ -120,6 +120,14 @@ StatefulSDPAFusion::StatefulSDPAFusion() {
         const auto& pattern_map = m.get_pattern_value_map();
         auto root = m.get_match_root();
 
+        // This fusion has no operand able to carry the mapping from quantization codes to values,
+        // and the node it builds would widen them as magnitudes. Decline, so the graph reaches
+        // ScaledDotProductAttentionDecomposition and is reported against the SDPA node by name.
+        const auto sdpa_root = ov::as_type_ptr<ov::op::v13::ScaledDotProductAttention>(root);
+        if (sdpa_root && ov::op::v13::ScaledDotProductAttention::has_quantized_kv(*sdpa_root)) {
+            return false;
```
[the remark this answers is not in the document]
Comment (dmenig) [reply src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/stateful_sdpa_fusion.cpp:128]: > Agreed, and fixed, using the harness you pointed at. `state_concat_sdpa.cpp` gains `StateConcatSDPAQuantizedKVShouldNotFuse`, which builds a stateful i8 K/V cache graph and registers `StatefulSDPAFusion` with `model_ref` set to the same unmodified graph — so the assertion is precisely that the original SDPA survives and no `ScaledDotProductAttentionWithKVCache` is constructed.
>
> Mutation-checked rather than asserted: disabling the guard fails this test and only this test, and the other 15 tests in the suite are unaffected either way.
>
> On the other four guards, plainly: the PagedAttention one is covered by the test added for your other comment. `SDPAFuseTransposeReshape`, the GPU decomposition callback and the template reference remain verified by hand. I would rather say that than imply the sweep is complete.
>
> Landed in b3b77e0628, as a separate commit on top of the reviewed head.

**Comment by p-durandin** [comment]
build_jenkins
